public/src/components/channelManagement/campaigns/StatusUpdateDialog.tsx

```diff
@@ -0,0 +1,165 @@
+import React, { useState, useEffect } from 'react';
+import {
+  Button,
+  Dialog,
+  DialogActions,
+  DialogContent,
+  DialogTitle,
+  List,
+  ListItem,
+  IconButton,
+  makeStyles,
+  Switch,
+  Typography,
+} from '@material-ui/core';
+import CloseIcon from '@material-ui/icons/Close';
+import { Test, Status } from '../helpers/shared';
+import { updateStatuses, FrontendSettingsType } from '../../../utils/requests';
+import { testChannelData } from './CampaignsEditor';
+
+const useStyles = makeStyles(() => ({
+  dialogHeader: {
+    display: 'flex',
+    alignItems: 'center',
+    justifyContent: 'space-between',
+    paddingRight: '8px',
+  },
+  input: {
+    '& input': {
+      textTransform: 'uppercase !important',
+    },
+  },
+  onOffLabel: {
+    fontSize: '13px',
+  },
+  channelLabel: {
+    fontSize: '14px',
+    marginLeft: '16px',
+  },
+  testLabel: {
+    fontSize: '14px',
+    marginLeft: '10px',
+  },
+}));
+
+interface StatusUpdateDialogProps {
+  isOpen: boolean;
+  close: () => void;
+  tests: Test[];
+  updatePage: () => void;
+}
+
+interface TestStatus {
+  [index: string]: string;
+}
+
+type UpdateStatusesArguments = [FrontendSettingsType, string[], Status];
+
+const StatusUpdateDialog: React.FC<StatusUpdateDialogProps> = ({
+  isOpen,
+  close,
+  tests,
+  updatePage,
+}: StatusUpdateDialogProps) => {
+  const classes = useStyles();
+
+  const oldTestStatuses: TestStatus = {};
+
+  const [testData, setTestData] = useState<TestStatus>(oldTestStatuses);
```

Comment:
Here we're initialising a `useState` with a value which is mutated elsewhere, which isn't very idiomatic and I think potentially buggy.
You can implement this without defining `oldTestStatuses`.

First, you can initialise a `useState` with a function (instead of the `useEffect` below):
```
const [testData, setTestData] = useState<TestStatus>(() => {
  const oldTestStatuses: TestStatus = {};
  tests.map(test => {
    const key = getStatusKey(test);
    oldTestStatuses[key] = test.status;
  });
  return oldTestStatuses;
});
```

Later, in the `onSubmit` function where you're iterating over `tests`, you can instead use:
`if (test.status !== testData[key])`

Reply:
I tried the solution, but it doesn't work. The campaign's test data is asynchronous and doesn't resolve until after the component completes loading (as part of the page load).

This code works. Does it address your mutated object worries?

```
  const [testData, setTestData] = useState<TestStatus>({});

  useEffect(() => {
    const res: TestStatus = {};
    tests.map(test => {
      const key = getStatusKey(test);
      res[key] = test.status;
    });
    setTestData(res);
  }, [tests]);

```